compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -222,14 +222,21 @@ pub struct MethodDef<'a> {
 
     pub attributes: ast::AttrVec,
 
-    /// Can we combine fieldless variants for enums into a single match arm?
-    /// If true, indicates that the trait operation uses the enum tag in some
-    /// way.
-    pub unify_fieldless_variants: bool,
+    pub unify_fieldless_variants: UnifyFieldlessVariants,
 
     pub combine_substructure: RefCell<CombineSubstructureFunc<'a>>,
 }
 
+/// Can we combine fieldless variants for enums into a single match arm?
+#[derive(PartialEq)]
+pub enum UnifyFieldlessVariants {
```

Comment:
I'd suggest the following:
```rust
/// How to handle fieldless enum variants.
#[derive(PartialEq)]
pub enum FieldlessVariantsStrategy {
    /// Combine fieldless variants into a single match arm.
    /// This assumes that relevant information has been handled
    /// by looking at the enum's discriminant.
    Unify,

    /// Don't do anything special about fieldless variants. They are
    /// handled like any other variant.
    Default,

    /// If all variants of the enum are fieldless, expand the special
    /// `AllFieldLessEnum` substructure, so that the entire enum can be handled
    /// at once.
    SpecializeIfAllVariantsFieldless,
}
```